// App/Module.js
var Class = require('jn-classjs'),
//	Cache = require('App/Cache'),
	Config = require('App/Config');

/**
 * @class App/Module
 */
module.exports = Class.create({
	extend: require('events').EventEmitter,
	constructor: function(app, name, path, env) {
		this.app = app;
		this.env = env;
		this.name = name;
		this.path = path;
		this.initConfig();
	},

	/*getController: function(name) {
		var realName = getControllerClass(name);

		if (!this.controllers.has(realName)) {
			try {
				var ControllerClass = require(this.path + '/controllers/' + realName);
			} catch (e) {
				this.throwDispatchException(err.controllerNotFound);
			}
			var object = new ControllerClass(this.application);

			this.controllers.set(realName, object, 3600000);	// 1h
		} else {
			object = this.controllers.keepAlive(realName);
		}

		return object;
	},*/

	initConfig: function() {
		var config = new Config(this.path + '/config.json');
		this.config = config.getSection(this.env);
	},

	getApplication: function() {
		return this.app;
	}
});
