// App/Router/Route.js
var Class = require('jn-classjs');

module.exports = Class.create({
	defaults: null,
	params: null,

	constructor: function(data) {
		this.defaults = data.defaults || {};
		this.module = data.module;

		var action = data.action.replace('-', '/').split(':');
		if (action.length !== 2) {
			throw new Error('Invalid route action');
		}

		this.controller = action[0];
		this.action = action[1];

		var methods = data.methods || 'get';
		if (typeof methods === 'string') {
			methods = methods.toLowerCase().split('|');
		} else {
			methods = methods.join('|').toLowerCase().split('|');
		}
		this.methods = methods;

		this.url = data.url || '/';
		this.layout = data.layout || 'layout';
	},

	getLayout: function() {
		return this.layout;
	},

	getModule: function() {
		return this.module;
	},

	getController: function() {
		return this.controller;
	},

	getAction: function() {
		return this.action;
	},

	getDefaults: function() {
		return this.defaults;
	},

	getMethods: function() {
		return this.methods;
	},

	getUrl: function() {
		return this.url;
	}
});
